{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "65bffa72",
   "metadata": {},
   "source": [
    "# Combining and Filtering Data Tables in `pandas` "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a45e3393",
   "metadata": {},
   "source": [
    "## In this section\n",
    "\n",
    "- Using the `merge` and `join` methods to combine pandas `DataFrames`\n",
    "- Filtering missing data out of pandas `DataFrames`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f63c666",
   "metadata": {},
   "source": [
    "## Prerequisites\n",
    "\n",
    "- Running python\n",
    "- Basic variable assignment and calling functions in python\n",
    "- Loading pandas `DataFrames`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8441d76",
   "metadata": {},
   "source": [
    "## Table of Content\n",
    "\n",
    "* [Bioinformatic analysis often involves combining data from multiple sources](#Bioinformatic-analysis-often-involves-combining-data-from-multiple-sources)\n",
    "* [Merging and Joining `pandas` `DataFrames`](#Merging-Tables-in-Pandas)\n",
    "* [Handling Missing Data](#Handling-Missing-Data)\n",
    "   \n",
    "   -[Method 1: the `dropna` method](#Method-1---use-the-built-in-dropna-method)\n",
    "   \n",
    "   -[Method 2: identify and filter out `NaN` rows](#Method-2---identify-and-remove-rows-with-NaN-values-using-row-filtering-and-logic-operations)\n",
    "* [A Biological Example: Combining Coral Datasets](#A-Biological-Example---Combining-Coral-Datasets)\n",
    "   \n",
    "* [Exercises](#Exercises)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1aceed15",
   "metadata": {},
   "source": [
    "## Bioinformatic analysis often involves combining data from multiple sources\n",
    "\n",
    "In most interesting bioinformatic analyses, the data we want to compare doesn't all start in one place. For example, if we were studying how aging affects the microbiome, perhaps we have one table that tells us how old each patient is, and another that contains a summary of microbiome diversity data for each patient. \n",
    "\n",
    "If there are hundreds or thousands of patients, the order of patients differs between datasets, or if some data are missing (one patient didn't want to give her age, one microbiome amplicon library failed to sequence), it can be very obnoxious to match up these datasets by hand. Worse, it would be easy to make a mistake that might influence your results.\n",
    "\n",
    "Luckily, `pandas` `DataFrame` objects provide good methods for combining data. Additionally, they include useful tools for filtering your dataset to remove missing data. Both these activities often end up being important parts of larger bioinformatics projects, so it's worth devoting some time to practicing them."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "423aec25",
   "metadata": {},
   "source": [
    "# Merging Tables in Pandas\n",
    "\n",
    "Let's walk through combining data from two tables, using a simple example. In this example, we have one file, `gene_id_vs_length.txt` that says the length of many genes in a genome. (**Side note:**. You could make such a file yourself by using a FASTA parser to parse a genome file, a function you wrote to calculate the length of each gene, and then writing out the results into a `.tsv` file by separating out each gene id from each length with a tab or '\\t' character).\n",
    "\n",
    "You have a second file `gene_id_vs_hgt.txt` that says whether there is evidence that each gene underwent horizontal gene transfer (HGT) - a process in which unrelated species can exchange genetic material \"horizontally\" across the branches of the tree of life. (**Side note**: this process is super important for understanding the spread of antibiotic resistance genes, as these are commonly horizontally transferred between bacterial species).\n",
    "\n",
    "We want to compare whether horizontally transferred genes tend to be shorter, longer, or the same length as non-horizontally transferred genes. Therefore, we need to combine the information in our two tables into a single `pandas` `DataFrame`."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88990e32",
   "metadata": {},
   "source": [
    "## Download some example data\n",
    "\n",
    "First, let's download our data. In a real research project, you could either use code like that below to download data from an ftp site, or download it by hand and put it into the same folder that you ran your jupyter notebook from."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e7f8aaeb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading gene_id_vs_length.txt\n",
      "Downloading gene_id_vs_hgt.txt\n"
     ]
    }
   ],
   "source": [
    "#Import the urlretrieve function\n",
    "from urllib.request import urlretrieve\n",
    "\n",
    "#Save the looooong web address (url) of our data in a python string \n",
    "chapter_url = 'https://raw.githubusercontent.com/zaneveld/full_spectrum_bioinformatics/master/content/07_tabular_omics_data/resources/'\n",
    "#Set the name of the file where we want to save the data\n",
    "file_names = ['gene_id_vs_length.txt','gene_id_vs_hgt.txt']\n",
    "\n",
    "#Download the data \n",
    "for file_name in file_names:\n",
    "    print(f\"Downloading {file_name}\")\n",
    "    urlretrieve(chapter_url+file_name, file_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22906cda",
   "metadata": {},
   "source": [
    "## Check that we have the data files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2e2374e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['scenario_1_otus_pandas.tsv',\n",
       " 'rarefaction_5000.png',\n",
       " 'gene_id_vs_length.txt',\n",
       " 'rarefaction_10000.png',\n",
       " '.DS_Store',\n",
       " 'merging_tables.ipynb',\n",
       " 'analyzing_tabular_omics_data_in_pandas.ipynb',\n",
       " 'resources',\n",
       " 'rarefaction_1000.png',\n",
       " 'scenario_1_otus_pandas.txt',\n",
       " 'simulating_microbiome_sampling_depth.ipynb',\n",
       " 'tabular_omics_data.ipynb',\n",
       " '.ipynb_checkpoints',\n",
       " 'rarefaction_50.png',\n",
       " 'gene_id_vs_hgt.txt',\n",
       " 'rarefaction_500.png']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from os import listdir\n",
    "listdir('./')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3844bc6",
   "metadata": {},
   "source": [
    "Check the files listed, and just make sure that 'gene_id_vs_length.txt', and 'gene_id_vs_hgt.txt' are in the list."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc6bb50f",
   "metadata": {},
   "source": [
    "## Load each table into a `pandas` `DataFrame`.\n",
    "\n",
    "Next, we'll import pandas and load each of our data tables into a pandas dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "de077b51",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>gene1</th>\n",
       "      <td>1200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene2</th>\n",
       "      <td>500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene3</th>\n",
       "      <td>750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene4</th>\n",
       "      <td>2000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene5</th>\n",
       "      <td>3500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene6</th>\n",
       "      <td>800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene7</th>\n",
       "      <td>900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene8</th>\n",
       "      <td>3000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene9</th>\n",
       "      <td>3200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene10</th>\n",
       "      <td>2800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene11</th>\n",
       "      <td>5300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene12</th>\n",
       "      <td>4000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         gene_length\n",
       "gene_id             \n",
       "gene1           1200\n",
       "gene2            500\n",
       "gene3            750\n",
       "gene4           2000\n",
       "gene5           3500\n",
       "gene6            800\n",
       "gene7            900\n",
       "gene8           3000\n",
       "gene9           3200\n",
       "gene10          2800\n",
       "gene11          5300\n",
       "gene12          4000"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "gene_length_df = pd.read_csv('./gene_id_vs_length.txt',sep = '\\t')\n",
    "gene_length_df = gene_length_df.set_index('gene_id')\n",
    "\n",
    "gene_length_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "83092604",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>horizontally_transferred</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>gene1</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene2</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene3</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene4</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene5</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene6</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene7</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene8</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene9</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene10</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene11</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene12</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         horizontally_transferred\n",
       "gene_id                          \n",
       "gene1                           0\n",
       "gene2                           1\n",
       "gene3                           1\n",
       "gene4                           0\n",
       "gene5                           0\n",
       "gene6                           1\n",
       "gene7                           1\n",
       "gene8                           0\n",
       "gene9                           0\n",
       "gene10                          1\n",
       "gene11                          0\n",
       "gene12                          0"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_hgt_df = pd.read_csv('./gene_id_vs_hgt.txt',sep = '\\t')\n",
    "gene_hgt_df = gene_hgt_df.set_index('gene_id')\n",
    "\n",
    "gene_hgt_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85c38ded",
   "metadata": {},
   "source": [
    "## Types of Joins \n",
    "\n",
    "There are several methods for joining tables together using Pandas DataFrames. Join methods differ in how they handle missing data. For example, if a gene is in your first table but not your second one, should it be included or excluded from the joined table? \n",
    "\n",
    "This type of problem comes up all the time in databases (essentially large collections of tables, linked together by shared ids). Therefore, folks who work with database languages like `SQL` have developed some vocabulary for describing different types of joins. It's worth learning the basics even if you don't use databases, because the same language is used in many other settings, including in `pandas`.\n",
    "\n",
    "Here are the most common ways of joining two tables together:\n",
    "\n",
    "**inner join** -- an inner join is the most strict type of joining or merging tables. It only joins rows that   that match for BOTH tables, and excludes all others.\n",
    "\n",
    "**outer join** -- an outer join is the least strict, most loosy-goosy join. An outer join take all rows that appear in *either* table, leave empty entries if data is missing\n",
    "\n",
    "**left join** -- data in the rightmost table is included only if it matches an id in the left table, otherwise it is excluded.\n",
    "\n",
    "**right join** -- data in the leftmost table is included only if it matches an id in the left table, otherwise it is excluded."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c26b3cad",
   "metadata": {},
   "source": [
    "## Combining tables with `join`\n",
    "We will use an 'outer' join to put together our two tables. Sensibly enough, there is a DataFrame method called `join` that will join your DataFrame to another."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "efb2e04b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_length</th>\n",
       "      <th>horizontally_transferred</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>gene1</th>\n",
       "      <td>1200</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene2</th>\n",
       "      <td>500</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene3</th>\n",
       "      <td>750</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene4</th>\n",
       "      <td>2000</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene5</th>\n",
       "      <td>3500</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene6</th>\n",
       "      <td>800</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene7</th>\n",
       "      <td>900</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene8</th>\n",
       "      <td>3000</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene9</th>\n",
       "      <td>3200</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene10</th>\n",
       "      <td>2800</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene11</th>\n",
       "      <td>5300</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene12</th>\n",
       "      <td>4000</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         gene_length  horizontally_transferred\n",
       "gene_id                                       \n",
       "gene1           1200                         0\n",
       "gene2            500                         1\n",
       "gene3            750                         1\n",
       "gene4           2000                         0\n",
       "gene5           3500                         0\n",
       "gene6            800                         1\n",
       "gene7            900                         1\n",
       "gene8           3000                         0\n",
       "gene9           3200                         0\n",
       "gene10          2800                         1\n",
       "gene11          5300                         0\n",
       "gene12          4000                         0"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df = gene_length_df.join(gene_hgt_df, how = 'outer')\n",
    "\n",
    "combined_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8bfbf9c",
   "metadata": {},
   "source": [
    "## Handling Missing Data\n",
    "\n",
    "It is common to have missing data in Bioinformatics dataset. Even if each of your data sources is 100% complete, they may cover different sets of species, resulting in missing data when you combine them. For example, you might have physiological but not genomic data for one organism, and genomic but not physiological data for another.\n",
    "Unfortunately, organisms lacking either type of data would normally have to be removed from your analysis.\n",
    "\n",
    "Inside `pandas`, missing data is shown as `NaN`, short for 'not a number'. So when we have missing data, we often have to do some work to filter out any rows of our `pandas` `DataFrame` that have `NaN` values.\n",
    "\n",
    "To explore this, let's artificially add missing data (and therefore `NaN` values) to our dataframe. You would never do this in practice, of course, but it will provide us a simple example to work with, where we know exactly which data we want to exclude."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49fb264d",
   "metadata": {},
   "source": [
    "### Artificially introduce 2 missing data points to our table, as an example\n",
    "\n",
    "We will use the `DataFrame` `.loc` property to set the value of gene length to `None` for gene10, and the value of horizontally_transferred to `None` for gene7. Recall that the `.loc` property lets you specify entries in the `DataFrame` based on the name of the row, then the column, both of which we right inside square brackets (similar to indexing into a list or retrieving a key from a dict). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "id": "f83604a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Artificially introduce some missing data\n",
    "\n",
    "combined_df.loc['gene10','gene_length'] = None\n",
    "combined_df.loc['gene7','horizontally_transferred'] = None"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa5ac358",
   "metadata": {},
   "source": [
    "### Examine our dataframe to confirm that some data are now missing\n",
    "\n",
    "Next let's take a look at our dataframe so we can see what this missing data looks like."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "id": "b84a66b9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_length</th>\n",
       "      <th>horizontally_transferred</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>gene1</th>\n",
       "      <td>1200.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene2</th>\n",
       "      <td>500.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene3</th>\n",
       "      <td>750.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene4</th>\n",
       "      <td>2000.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene5</th>\n",
       "      <td>3500.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene6</th>\n",
       "      <td>800.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene7</th>\n",
       "      <td>900.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene8</th>\n",
       "      <td>3000.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene9</th>\n",
       "      <td>3200.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene10</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene11</th>\n",
       "      <td>5300.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene12</th>\n",
       "      <td>4000.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         gene_length  horizontally_transferred\n",
       "gene_id                                       \n",
       "gene1         1200.0                       0.0\n",
       "gene2          500.0                       1.0\n",
       "gene3          750.0                       1.0\n",
       "gene4         2000.0                       0.0\n",
       "gene5         3500.0                       0.0\n",
       "gene6          800.0                       1.0\n",
       "gene7          900.0                       NaN\n",
       "gene8         3000.0                       0.0\n",
       "gene9         3200.0                       0.0\n",
       "gene10           NaN                       1.0\n",
       "gene11        5300.0                       0.0\n",
       "gene12        4000.0                       0.0"
      ]
     },
     "execution_count": 202,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b081b379",
   "metadata": {},
   "source": [
    "As expected we nbow see 'NaN' written in the two places where we have missing data."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28dd90d1",
   "metadata": {},
   "source": [
    "## Method 1 - use the built-in dropna method\n",
    "\n",
    "The first way we can filter out rows that have missing data is to use the built in `.dropna()` method of `pandas` `DataFrames`. This will filter out *all* rows that have *any* missing data in *any* column, and return the resulting DataFrame to us. \n",
    "\n",
    "For simple cases like this one, `.dropna()` is perfect. Here's how it works:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "id": "b94b1fe9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_length</th>\n",
       "      <th>horizontally_transferred</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>gene1</th>\n",
       "      <td>1200.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene2</th>\n",
       "      <td>500.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene3</th>\n",
       "      <td>750.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene4</th>\n",
       "      <td>2000.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene5</th>\n",
       "      <td>3500.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene6</th>\n",
       "      <td>800.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene8</th>\n",
       "      <td>3000.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene9</th>\n",
       "      <td>3200.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene11</th>\n",
       "      <td>5300.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene12</th>\n",
       "      <td>4000.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         gene_length  horizontally_transferred\n",
       "gene_id                                       \n",
       "gene1         1200.0                       0.0\n",
       "gene2          500.0                       1.0\n",
       "gene3          750.0                       1.0\n",
       "gene4         2000.0                       0.0\n",
       "gene5         3500.0                       0.0\n",
       "gene6          800.0                       1.0\n",
       "gene8         3000.0                       0.0\n",
       "gene9         3200.0                       0.0\n",
       "gene11        5300.0                       0.0\n",
       "gene12        4000.0                       0.0"
      ]
     },
     "execution_count": 204,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_df = combined_df.dropna()\n",
    "filtered_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0981604",
   "metadata": {},
   "source": [
    "Great! The `.dropna()` method has removed rows with missing data, so gene 10 and gene 7 are now excuded from our table."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dfc93486",
   "metadata": {},
   "source": [
    "## Method 2 - identify and remove rows with `NaN` values using row filtering and logic operations\n",
    "\n",
    "For most practical uses, `.dropna` works great. But sometimes we want more control over what gets dropped. Or perhaps we want to save all the rows with problems so we can look at them by hand - usally a very important step to make sure everything we're removing is justified.\n",
    "\n",
    "Implementing a similar process by hand is also a very useful way to understand how `pandas` `DataFrame` objects work, since we will break up a complex process into smaller steps, each of which can be examined one by one."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf144bbf",
   "metadata": {},
   "source": [
    "### Building a DataFrame of Boolean values showing which table entries have `NaN` values \n",
    "\n",
    "First, we'll use the `isnull()` function to build a table with `True` every place that the data table has an `NaN`, and `False` everywhere else.\n",
    "\n",
    "*If our method works, we expect `False` everywhere except the gene length entry for `gene10` and the `horizontally_transferred` column for `gene7`.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "id": "de50059f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         gene_length  horizontally_transferred\n",
      "gene_id                                       \n",
      "gene1          False                     False\n",
      "gene2          False                     False\n",
      "gene3          False                     False\n",
      "gene4          False                     False\n",
      "gene5          False                     False\n",
      "gene6          False                     False\n",
      "gene7          False                      True\n",
      "gene8          False                     False\n",
      "gene9          False                     False\n",
      "gene10          True                     False\n",
      "gene11         False                     False\n",
      "gene12         False                     False\n"
     ]
    }
   ],
   "source": [
    "bad_values = pd.isnull(combined_df)\n",
    "print(bad_values)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36859c3a",
   "metadata": {},
   "source": [
    "## Find bad rows from table of null values\n",
    "\n",
    "Second, we want to figure out which rows are no good using our giant table of all the individual entries in our table that are `NaN`. \n",
    "\n",
    "We now have a whole dataframe filled with True or False, based on which entries are `NaN`. However, to be able to select or filter rows, we need a single column of True or False values. Depending on the application, we might throw out rows that have any `NaN` values (i.e. `NaN` in at least one column), or only those that have all `NaN` values (`NaN` in every column).\n",
    "\n",
    "In pandas there are DataFrame methods that can automate the process of finding rows where `any` or `all` columns meet some criterion. These methods can then be applied to a DataFrame full of `True` or `False` values to find our bad rows.\n",
    "\n",
    "These methods work for both rows or columns, so we need to specify `axis = 1` (columns) to get the result we expect. The alternative setting (`axis = 0`, the default value) would be more useful if we were getting rid of *columns* that had an `NaN` in any row.\n",
    "\n",
    "**If our method works, we expect `False` for all genes except `gene7` and `gene10` where we added `NaN` values**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "id": "0cbbf6ce",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene_id\n",
      "gene1     False\n",
      "gene2     False\n",
      "gene3     False\n",
      "gene4     False\n",
      "gene5     False\n",
      "gene6     False\n",
      "gene7      True\n",
      "gene8     False\n",
      "gene9     False\n",
      "gene10     True\n",
      "gene11    False\n",
      "gene12    False\n",
      "dtype: bool\n"
     ]
    }
   ],
   "source": [
    "bad_rows = bad_values.any(axis=1)\n",
    "print(bad_rows)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8bb69da",
   "metadata": {},
   "source": [
    "### Filter the table to include all the rows except the bad ones\n",
    "\n",
    "Our series of True and False values can now be used to select rows from our original data table using regular row indexing.\n",
    "\n",
    "One big help here will be the `~` (tilda) character, which inverts `True` and `False` in selections in pandas. This will let us select all the good rows based on their having `False` values in the bad_rows Series."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "id": "dcb271b7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         gene_length  horizontally_transferred\n",
      "gene_id                                       \n",
      "gene1         1200.0                       0.0\n",
      "gene2          500.0                       1.0\n",
      "gene3          750.0                       1.0\n",
      "gene4         2000.0                       0.0\n",
      "gene5         3500.0                       0.0\n",
      "gene6          800.0                       1.0\n",
      "gene8         3000.0                       0.0\n",
      "gene9         3200.0                       0.0\n",
      "gene11        5300.0                       0.0\n",
      "gene12        4000.0                       0.0\n"
     ]
    }
   ],
   "source": [
    "filtered_df = combined_df[~bad_rows]\n",
    "print(filtered_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e7834fd",
   "metadata": {},
   "source": [
    "### Retain bad rows for later inspection.\n",
    "\n",
    "It's a good idea to keep a copy of the bad rows so you can figure out what went wrong. We can do this by repeating our command without inverting the selection with `~`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "id": "48141d6f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         gene_length  horizontally_transferred\n",
      "gene_id                                       \n",
      "gene7          900.0                       NaN\n",
      "gene10           NaN                       1.0\n"
     ]
    }
   ],
   "source": [
    "bad_rows_df = combined_df[bad_rows]\n",
    "print(bad_rows_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1191170a",
   "metadata": {},
   "source": [
    "# A Biological Example - Combining Coral Datasets\n",
    "\n",
    "Let's practice what we've learned about joining and filtering `pandas` `DataFrames` to join together three tables with information on corals.\n",
    "\n",
    "We want to put together a table that will help us figure out if the microbes that live on corals are correlated with how often they get disease. We have three tables, that each hold different important pieces of information about disease or microbial symbiosis in several different kinds of corals. Each of these tables is a `.tsv` file, meaning that `\\t` characters separate the columns of each dataframe.\n",
    "\n",
    "`coral_disease_susceptibility.txt` - this file says the average prevalence of disease in diver surverys of each type of coral. \n",
    "\n",
    "`coral_microbiome diversity.txt` - this file says how diverse each coral's microbiome is on average, based on 16S rRNA gene amplicon sequencing.\n",
    "\n",
    "`coral_symbiont_abundance.txt` - this file says how common a particular bacterium, *Endozoicomonas* was in 16S rRNA gene amplicon sequencing from that species.\n",
    "\n",
    "The common identifier that will let us link together the tables will be the genus name of corals (*Acropora*), *Porites*, etc). Let's first check that we have all the tables we need in our current folder, and then move on to loading the three dataframes."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf580735",
   "metadata": {},
   "source": [
    "## Download coral data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ed80cd5e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading coral_disease_susceptibility.txt\n",
      "Downloading coral_microbiome_diversity.txt\n",
      "Downloading coral_symbiont_abundance.txt\n"
     ]
    }
   ],
   "source": [
    "#Import the urlretrieve function\n",
    "from urllib.request import urlretrieve\n",
    "\n",
    "#Save the looooong web address (url) of our data in a python string \n",
    "chapter_url = 'https://raw.githubusercontent.com/zaneveld/full_spectrum_bioinformatics/master/content/07_tabular_omics_data/resources/'\n",
    "#Set the name of the file where we want to save the data\n",
    "file_names = ['coral_disease_susceptibility.txt', 'coral_microbiome_diversity.txt', 'coral_symbiont_abundance.txt']\n",
    "\n",
    "#Download the data \n",
    "for file_name in file_names:\n",
    "    print(f\"Downloading {file_name}\")\n",
    "    urlretrieve(chapter_url+file_name, file_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa8f09bd",
   "metadata": {},
   "source": [
    "## Load all three DataFrames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "5b452dcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "micro_div_df = pd.read_csv('coral_microbiome_diversity.txt',sep=\"\\t\")\n",
    "endo_df = pd.read_csv('coral_symbiont_abundance.txt',sep='\\t')\n",
    "disease_df = pd.read_csv('coral_disease_susceptibility.txt',sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b776876",
   "metadata": {},
   "source": [
    "## Set the host genus column to be the index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a3b15c47",
   "metadata": {},
   "outputs": [],
   "source": [
    "micro_div_df = micro_div_df.set_index(\"host_genus\")\n",
    "endo_df = endo_df.set_index(\"host_genus\")\n",
    "disease_df = disease_df.set_index(\"host_genus\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d86a9e73",
   "metadata": {},
   "source": [
    "## `join` tables using their indices\n",
    "\n",
    "When using the `join` method, we can only combine two tables at a time. So we'll have to pick any two of our tables to stick together to make a combined DataFrame, and then use a second `join` command to add the third table to that combined dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "763a8230",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>average_endozoicomonas_per_1000_reads</th>\n",
       "      <th>disease_susceptibility</th>\n",
       "      <th>faith_pd_tissue</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>host_genus</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Acanthastrea</th>\n",
       "      <td>8.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>62.672799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Acropora</th>\n",
       "      <td>58.681818</td>\n",
       "      <td>3.643113</td>\n",
       "      <td>22.585867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Aiptasia</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>132.477305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Alveopora</th>\n",
       "      <td>441.333333</td>\n",
       "      <td>0.749064</td>\n",
       "      <td>98.928105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Astrea</th>\n",
       "      <td>302.666667</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19.777494</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Stylophora</th>\n",
       "      <td>634.428571</td>\n",
       "      <td>3.687151</td>\n",
       "      <td>28.399778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tubastraea</th>\n",
       "      <td>7.285714</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>104.221069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tubipora</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Turbinaria</th>\n",
       "      <td>38.000000</td>\n",
       "      <td>1.872075</td>\n",
       "      <td>82.013983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Xenia</th>\n",
       "      <td>302.500000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25.127500</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>74 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              average_endozoicomonas_per_1000_reads  disease_susceptibility  \\\n",
       "host_genus                                                                    \n",
       "Acanthastrea                               8.000000                0.000000   \n",
       "Acropora                                  58.681818                3.643113   \n",
       "Aiptasia                                        NaN                     NaN   \n",
       "Alveopora                                441.333333                0.749064   \n",
       "Astrea                                   302.666667                     NaN   \n",
       "...                                             ...                     ...   \n",
       "Stylophora                               634.428571                3.687151   \n",
       "Tubastraea                                 7.285714                0.909091   \n",
       "Tubipora                                        NaN                     NaN   \n",
       "Turbinaria                                38.000000                1.872075   \n",
       "Xenia                                    302.500000                     NaN   \n",
       "\n",
       "              faith_pd_tissue  \n",
       "host_genus                     \n",
       "Acanthastrea        62.672799  \n",
       "Acropora            22.585867  \n",
       "Aiptasia           132.477305  \n",
       "Alveopora           98.928105  \n",
       "Astrea              19.777494  \n",
       "...                       ...  \n",
       "Stylophora          28.399778  \n",
       "Tubastraea         104.221069  \n",
       "Tubipora                  NaN  \n",
       "Turbinaria          82.013983  \n",
       "Xenia               25.127500  \n",
       "\n",
       "[74 rows x 3 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#First make a combined DataFrame by joining the disease and Endozoicomonas DataFrames\n",
    "combined_df = endo_df.join(disease_df,how='outer')\n",
    "\n",
    "#Now add in the microbial diversity data with another join statement\n",
    "combined_df = combined_df.join(micro_div_df,how='outer')\n",
    "\n",
    "#Look at our combined DataFrame\n",
    "combined_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "408f3dc3",
   "metadata": {},
   "source": [
    "## Drop rows with missing data\n",
    "\n",
    "As you may have noticed in the combined table, there are a lot of `NaN` values! We probably want to exclude missing data from our analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b7c54350",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_df = combined_df.dropna()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81ef731b",
   "metadata": {},
   "source": [
    "**Stop and check**: did this command remove all the `NaN` values from your table?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d5f5820",
   "metadata": {},
   "source": [
    "## Save our combined data to a file\n",
    "\n",
    "It's a good idea to save your final, processed dataset to a file. This helps ensure that anything you notice in a graph can be easily found in your data. If you publish your analysis in a scientific journal, the raw data for each figure should be made available (e.g. at many journals it can be put into an Excel spreadsheet and uploaded as a \"Supplementary Data\" file). \n",
    "\n",
    "In the past, I have had to use image analysis software to literally click on the points of a scatterplot to get the raw data values because they were not made available by the authors - and perhaps surprisingly this is not an uncommon occurence! We never want to put our readers through that, so uploading raw data is the way to go. \n",
    "\n",
    "We can use the pandas `.to_csv` method to save out combined DataFrame as a data file. If we want a tab-separated values or '.tsv' file, we'll set 'sep' to '\\t' (just like when we loaded the file), and make sure the filename ends in '.txt' or '.tsv'. If we want a comma separated values (.csv) file, we'll set sep = ',' and make sure the filename ends in '.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "8becbf65",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_df.to_csv(\"combined_coral_data.tsv\",sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79064199",
   "metadata": {},
   "source": [
    "## Make a scatterplot\n",
    "\n",
    "Finally, we can extract any columns we want into python lists, and use them for other analyses like making scatterplots or other graphs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "0b63b566",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0, 0.5, 'Disease Susceptibility')"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYYAAAEOCAYAAACNY7BQAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAtB0lEQVR4nO3de5xT1bn/8c/DRUFugiBakQGKiihaKPVeQaHeSkuvViwVEQ/+Tm2FU4v1crzVSy1IRbS1tcVbaVFbtajHVhxUrNpqUVRUVBAYRUFAEIebCPP8/lh7JAlJJptkksnM9/165ZXstVf2fpLJ5Mnea+21zN0RERGp1azUAYiISMOixCAiIkmUGEREJIkSg4iIJFFiEBGRJC1KHUAhdO7c2Xv06FHqMEREysoLL7yw2t27pJY3isTQo0cP5s6dW+owRETKiplVpSvXqSQREUmixCAiIkmUGEREJIkSg4iIJFFiEBGRJEoMIiIN1OzKSkaPGMbxR/Vl9IhhzK6sLMp+G0V3VRGRxmZ2ZSXTJo9jwvDW9OvZhflLVjBp8jjgRoYMHVqv+9YRg4hIAzR92hQmDG9N/97taNG8Gf17t2PC8NZMnzal3vetxCAi0gBVVS2mX882SWX9erahqmpxve9biUFEpAGqqOjF/CUbksrmL9lARUWvet+3EoOISAM0csx4Js3cxLxF1WzdVsO8RdVMmrmJkWPG1/u+1fgsItIAhQbmG5k6bQpVdy2moqIXY84fX+8Nz6DEICLSYA0ZOrQoiSCVTiWJiEgSJQYREUmSc2Iws3fM7Goz268+AxIRkdKKc8SwGLgIeMPMnjGzs82sfT3FJSIiJZJzYnD3wUBP4HKgM3ArsNzMpptZ8VtHRESkXsRqY3D3d9z9anc/ADgauAs4GXhUp5pERBqHnW58dvd/uft/AxXAPUA34GLCqaanzGxYgWIUEZEi2unEYGYDzewmYAlwGrAQuAS4EOgKzDSzy7I8f18ze8LMFpjZa2Y2LirvZGaPmdnC6L7jzsYoIiLxxUoMZvY5M7vAzF4DngPOBB4CjnX3A9z9F+4+CegD/An4UZbNbQXOd/cDgSOAc82sLyGxzHb3/YDZ0bKIiBRJzlc+m9k/gCFAc+Bp4GzgXnffkFrX3T2qPzLT9tx9ObA8elxtZguAfYDhwOCo2p3Ak8DPco1TRETyE2dIjIOBicDt7r4oh/qPAcflsmEz6wH0JxyFdI2SBu6+3Mz2zPCcscBYgO7du+eyGxERyUGcxLCvu3uuld19FTCnrnpm1ha4Dxjv7h+bWa7bv5XQZZaBAwfmHJeIiGQXp41hq5mNyLTSzL5nZtvi7NzMWhKSwp/c/f6o+AMz2ztavzewMs42RUQkP3ESg0W3bNvK+Ze7hUODacACd/9VwqoHgVHR41HAzBgxiohInuIOu53ti/9w4KMY2zoa+AEw38xeisouBq4D7jWzMcA7wHdjxigiInnImhiiawvGJRRNMbNr0lTdHegAzMh1x+7+NJmPQIbkuh0RESmsuo4YPgbeix73ANYCq1LqOPAG8B/g+kIGJyIixZc1Mbj77cDtAGa2BLjA3R8sRmAiIlIaObcxuHvP+gxEREQaBs3gJiIiSTIeMZhZDVAD7ObuW6LlurqjurvH7ekkIiINSLYv8Z8TEsHWlGUREWnEMiYGd78i27KIiDROamMQEZEk2doYjt2ZDbr7UzsfjoiIlFq2NoYnidemYFH95vkEJCIipZUtMeQ0l4KIiDQu2Rqf65xLQUREGh81PouISJJsjc+XEdoMrnH3mmi5Lu7uVxUsOhERKTrLNFtnwpXOrROufK6Lu3vRG58HDhzoc+fOLfZuRUTKmpm94O4DU8uztTE0y7YsIiKNk77sRUQkSewB78ysGTAAqB2GewnworvncqpJREQauFiJwcxOJczStk9tEaEd4j0zm+Du9xQ4PhERKbKcE4OZnQb8GagC/hd4M1rVBxgL/NnMmrl7zvM+i4hIwxPniOFS4BXgGHdfn7jCzKYCzxIShhKDiEgZi9P43Au4IzUpAERlt0V1RESkjMVJDEuB9lnWt4/qiIhIGYuTGK4DzjWzA1NXmNlBwLnALwoVmEhTMbuyktEjhnH8UX0ZPWIYsysrSx2SNHHZhsT4eZri94FXzOwx4C1Cj6Q+wFBgPvD5+ghSpLGaXVnJtMnjmDC8Nf16dmH+khVMmjwOuJEhQ4eWOjxpouoaEiMuDYkhEsPoEcM4b9AK+vdu91nZvEXVTJ2zF7fPeLiEkUlTEHtIDLZfwCYi9aSqajH9enZJKuvXsw1Vdy0uUUQi2cdKqipmICJNUUVFL+YvST5imL9kAxUV6uAnpaOxkkRKaOSY8UyauYl5i6rZuq2GeYuqmTRzEyPHjC91aNKEZWt8vo3QuDzW3bdFy3Vxdx9TsOhEGrnQwHwjU6dNoequxVRU9GLM+ePV8Cwlla2N4XighnBUsS1aTt9SvV1d60UkxZChQ5UIpEHJ1sbQI9uyiIg0Tjm3MZhZdzNrnWV9azPrXpiwRESkVOI0Pi8Bvpll/dejOiIiUsbiJAarY30L1MYgIlL24nZXTfvFb2YdgJOBVXlHJCIiJZU1MZjZ5Wa2zcy2EZLC9NrlxBuwBhgB/LUIMYuISD2qa6KeVwizthlwOmEyntR2BAfWA/8B/pjrjqPrIoYBK9394KjsCuC/2H7kcbG7P5LrNkVEJH9ZE4O7PwA8AGBm+wBXu/vsAu37DuBm4K6U8hvc/foC7UNERGLKeWpPdz+ukDt296fMrEchtykiIvmLM+czAGZ2IvA1to++ugR4yN0fLVBMPzKzM4C5wPnuvjZDHGOBsQDdu+vyCRGRQsk4H8MOFc12A+4l9D4yoDq6b0toZ3gU+I67b8x55+GI4eGENoauwOpoe1cBe7v7WXVtR/MxiIjEl2k+hjjdVScCpwA3APu4ewd3bw/sA0wBTorq7DR3/8Ddt7l7DfB74LB8ticiIvHFSQynAn9095+6+/LaQndf7u7nA38CvpdPMGa2d8LiN4FX89meiIjEF6eNoTXwryzrnwWG57oxM5sBDAY6m9ky4HJgsJl9gXAqaSlwToz4RESkAOIkhmeBI4HfZlh/RFQnJ+4+Ik3xtBjxiIhIPYhzKulHhF/010aNxEBoMDazXwCDojoiIlLG4hwxzCacTvoZ8DMzqyac8mkfrV8LPG6WNNaeu3tFIQIVEZHiiJMYFqPRU0VEGr04Vz4Prsc4RESkgYg77LaIiDRysRKDmbU0s7PM7E9m9piZ9Y/KO5rZGWbWrX7CFBGRYsn5VJKZ7Q5UAgOATUAroGO0eh1wDTAduKiwIYqISDHFOWK4BuhLmNu5BwlTfUZDWNxPGBZDRETKWJzE8A3gZnd/GKhJs34hoK6pIiJlLk5i2AN4q446rfKIRUREGoA4iWEZsH+W9UcCi/ILR0RESi1OYrgXOMfMDk0ocwAzG0EYWXVGAWMTEZESiHPl89XA8cBzhNnVHLgq6q10IGHk1cmFDlBERIor5yOGaGa2Y4HLCGMmbQa+SEgQlwJD3H1LfQQpIiLFE2vO5+iLfyJ5ztQmIiINl4bEkEZjdmUlo0cM4/ij+jJ6xDBmV1aWOiSRspRzYjCzn5lZxol4zOwZMzu/MGGJxDO7spJpk8dx3qAVzLq8C+cNWsG0yeOUHER2Qpwjhu8TGp4z+TcwMr9wRHbO9GlTmDC8Nf17t6NF82b0792OCcNbM33alFKHJlJ24iSGXsCCLOvfBD6fXzgiO6eqajH9erZJKuvXsw1VVYtLFJFI+YqTGLYBXbKs70zC+EkixVRR0Yv5SzYklc1fsoGKil4likikfMVJDC8C3zOzXVJXRGWnAS8VKC6RWEaOGc+kmZuYt6iardtqmLeomkkzNzFyzPhShyZSduJ0V70B+Bswy8yuAOYTrmE4BLgcOAj4boHjE8nJkKFDgRuZOm0KVXctpqKiF2POHx+Vi0gc5p77NM5m9lPgWqB5yqoa4BJ3L8n1DQMHDvS5c+eWYtciImXLzF5w94Gp5XEvcLvezP4CfBvoHRUvBO5396r8wxQRkVKLlRgAogTwq3qIRUREGoDYiSGRme0GnA50Ihw1aNhtEZEyF+fK51vM7NWE5RbAM8DvgOuAl1KG5BYRkTIUp7vqccBDCcvfBQ4FfgQcBawCLilcaCIiUgpxTiXtAyReRjoMeNXdbwEws98CPy5gbCIiUgJxjhi2Ai0TlgcDsxKWVxPmhRYRkTIWJzG8AXzbgq8CewGPJqzvDqwpZHAiIlJ8cU4lXU+Y93kN0AZ4DXg8Yf0QYF7hQhMRkVLIOTG4+31mdgLwVWAd8Bt33wZgZp2AlcAf6yVKEREpmrhXPs8GZqcpXwN8q1BBiYhI6WhqTxERSZLzEYOZPV53Ldzdh+QRj4iIlFicU0nNCMNsJ2oO9CRc47AIeK9AcYmISInEaXwenGmdmZ0KTCWMupoTM7uNcJHcSnc/OCrrBNwD9ACWAqe6+9pctykiIvkrSBuDu99L6Mp6Q4yn3QGclFJ2ITDb3fcjNHJfWIj4REQkd4VsfH4VOCLXyu7+FDteEDccuDN6fCfwjYJEJiIiOStkYjgG2JjnNrq6+3KA6H7PTBXNbKyZzTWzuatWrcpztyIiUitOr6QzMqzqCBwPfA24qRBB5cLdbwVuhTC1Z7H2KyLS2MXplXRHlnUrgSsJ80Hn4wMz29vdl5vZ3tF2RUSkiOIkhp5pyhxY4+7rCxTPg8AowsQ/o4CZBdquiIjkKE531apC7tjMZhCG7u5sZsuAywkJ4V4zGwO8Q5gMSEREimin53w2s5bAYYSL2xa4+/w4z3f3ERlW6cppEZESytorycxONLPbzGyvlPL9CUNsPwXMIMz3PN3MNPaSiEiZq+uLfBTwZXdfkVI+HegL/BkYR5jJbQRwTsEjFBGRoqorMQwEHkosMLMvROX3ufsP3P1mdz8ZeB4YWS9RiohI0dSVGPYCFqaUDSL0RkqdlOd+wlGEiIiUsboSQ02aOl+K7p9OKV8F7FaIoEREpHTqSgxLgCNrF8ysOaHX0FvRrG2JOhOSg4iIlLG6uqveDVxlZs8DTwKjCeMX/S5N3aOAtwsanYiIFF1dieEm4DvAFEK7ggHzgV8lVjKzPYGTgasLH6KIiBRT1sTg7hvN7EjC8NefJ5xa+pu7b0mpujfwv8Bf6iNIEREpnjqvfHb3rcBf66jzMvByoYISEZHS0ZXKIiKSRIlBRESSKDGIiEgSJQYREUmixCAiIkmUGEREJEmsxGBmvczsLjN7z8y2mNnxUXnnaN6Gw+onTBERKZacE4OZHQjMBYYDC4DmtevcfTVwKJqPQUSk7MWZ2vMXwHrgCGALsDJl/d/RHM0iImUvzqmkQcCv3f19wrhJqaqAzxUkKhERKZk4iWFXYF2W9bsT5m8QEZEyFicxLCAMrZ3J19B4SSIiZS9OYvgtcLqZ/TewS1TmZtbFzG4BjgZ+XegARUSkuHJufHb335vZQYQv/9r5GB4kTOdpwK/c/Z7ChygiIsUUp1cS7j7ezGYA3wMOIBxxLAJmuPuz9RCfiIgUWazEAODuzwHP1UMsIiLSAMRODInMzIAhQCfgMXdfW5CoRESkZOJc+XyFmT2eUvwI8ChwN/C6mfUsZHAiIlJ8cXolfQt4qXbBzL4GnAhMBE4nDJHxv4UMTkREii/OqaTuwJsJy98A3nb3iwDMrA9wZsEik5KZXVnJ9GlTqKpaTEVFL0aOGc+QoUNLHZaIFEmcI4bmKcuDCaeRar0DdM03oIZkdmUlo0cM4/ij+jJ6xDBmV1aWOqR6N7uykmmTx3HeoBXMurwL5w1awbTJ45rEaxeRIE5ieBs4AcDMDgd6ALMS1u9D9iEzykpT/YKcPm0KE4a3pn/vdrRo3oz+vdsxYXhrpk+bUurQRKRI4iSGW4Bvmtl8wkiq75J8xHA08GoBYyuppvoFWVW1mH492ySV9evZhqqqxSWKSESKLefE4O6/A8YAC4GZwInu/gmAmXUCugAz6iPIUmiqX5AVFb2Yv2RDUtn8JRuoqOhVoohEpNhizeDm7re7+7fcfbS7v5lQvsbdv+ju0wofYmk01S/IkWPGM2nmJuYtqmbrthrmLapm0sxNjBwzvtShiUiRaM7nDOrrC7KhN2gPGTqUMeffyNQ5e3HClauYOmcvxpx/o3oliTQh5p5uzp0Mlc1aAt8EvgR0ZMfE4u4+Ju+gzJYC1cA2YKu7D8xWf+DAgT537tx8d7uDQnfbrG3QnjC8Nf16tmH+kg1MmrlJX7wiUhJm9kK679ecE4OZdQUeBw4EPgI6AGvYniBWA+vdPe9zLVFiGBjNJV2n+koMhTZ6xDDOG7SC/r3bfVY2b1E1U+fsxe0zHi5hZCLSFGVKDHFOJV0HdCNM8bk/Yajt7wFtgSuBDcBx+YfaeDXVBm0RKS9xEsNJwK3u/k+2z/ls7r7Z3a8E5gLXFyguB2aZ2QtmNjZdBTMba2ZzzWzuqlWrCrTb+tVUG7RFpLzESQwdgTeix1ui+90S1j8OHF+IoICj3X0AcDJwrpkdm1rB3W9194HuPrBLly4F2m39Uo8fESkHccZKWgnsCeDu1Wa2HuidsL49Ow6bsVPc/f3ofqWZPQAcBjxViG2XUmhgvpGp06ZQdVdo0B5zvsYhEpGGJU5ieAE4MmG5EhhnZnMJCeHHhNNJeTGzNkCzKPm0IQzD8fN8t5tJsQeMGzJ0qBKBiDRocRLDbcBZZtba3TcBPyP8in+S0BC9CvhpAWLqCjwQ5gCiBfBnd/9HAba7g+Tuo12Yv2QFkyaPA9R9VESarljXMezwZLO2hHaFbcAz7v5RgeKKZWe7q6r7qIg0ZZm6q+Y1tae7rwcezGcbpRS6jyY3XPfr2Yaqu9R9VESarlhDYphZSzM7y8z+ZGaPmVn/qLyjmZ1hZt3qJ8z6oe6jIiI7ijPn8+7Av4A/EGZvO57QhRXCPAzXAOcWNrz6pe6jIiI7inMq6RqgL/B14Dngg9oV7l5jZvcTLoK7qKAR1qPE7qPzbnqFZts203KX7XMuqAFaRBqq+uxRGedU0jeAm939YaAmzfqFQEUhgiqmIUOHMnLMePp2b8e08d15ZlKPepmtbdLEiQzosyef33tXBvTZk0kTJxZs2yLStNT3DJNxEsMewFt11GmVRywlU9+ztU2aOJGH7ryMqd9fz4JftWLq99fz0J2XKTmIyE6p7++sOIlhGWHwvEyOBBblF05p1PfgdjNuu55rT2vGMX12ZZcW4f7a05ox47ZCDS0lIk1JfX9nxUkM9wLnmNmhCWUOYGYjCCOtluXUnom9k9auXcuC11/hL488z0drVhfk0GzdunUc1rtlUtlhvVuybt26vLctIk1PffeojJMYrgYWEBqeZxKSwlVm9iowHfg3MLkgURVZbe+kOS++xztL3+LDdZt4cF4zxp7UoSDn7Tp06MDziz5NKnt+0ad06NAhr+3Wh4Y+w5yI1H+Pypx7Jbn7xmiU0/GEo4PNwBcJp48uBSa7+5bMW2i4ansn/c+PTqeVOQd0242zh3VjyIBOHN6nmqnTpuTV2j/irJ9y8Z2Xce1pn3BY75Y8v+hTLr67hhFnFWIEkcLRECGSSbHHFJPs6ntAzryGxGgo8p3BrfZD/9STj3HMQW0544SQFAC2bqvhhCtX8fizr+cV46SJE5lx2/WsW7eODh06MOKsnzLhggvy2mahaYgQSUdT0jZe9TIkRrThXYEO7r4y322VQuKH/ieHtWHd+k385uFFQG+GDOhUsPN2Ey64oMElglQaIkTSSewBA0Q9YMj7SFoarjhXPn/HzCanlF0KVAPLzWxWNEx2WUn80Hfr1p2ObYwffsW5a9ayJncltIYIkXQ0JW3TE6fx+X+ATrULZvZF4ArgGeD3hPmeJxQyuGJI/NB37NiRz+27Hx3atuLp19Yzdc5eTepwWUOESDr6wdD0xDmVtB/wl4Tl7wFrgJPdfbOZbY3KrihcePUvfOi3n1fv2LEjSz9swbGDm955dc0wJ+mMHDOeSZPHMWE4KW0M40sdmtSTOImhHfBxwvIQYJa7b46WXwTOLFBcRaMPfTLNMCep9IOh6YmTGN4DDgaIhtf+AnBzwvpOwCcFi6xI9KEXqZt+MDQtcRLDfYQ5nlsChwObgIcS1h8KlGVrlD70IiLbxWl8vhK4B/g+0AU4091XA5hZe+CbgC6TjegKYhEpV7GufAZGZVi9HtgH2FiIoIqpPq7o1BXEIlLO8r7ADcJEPYRZ3MpKIb7AUxNL3/7HMuO265lwYjWttuxG9cf70r93R10QJCJlI2NiMLMzood/dHdPWM7K3e8qSGRFkO8VnamJ5ZmX3+LSOyv5ZGMN3z6iNZ98uoV33l0I7Ee/nh10BbGIlIVsRwx3EEZQvRvYEi3XxYGySQz5DgGRmlj2bPUh157WjPNur+G1d2vo37MF3TtvZdnyd1n6YQtdECQiZSFbYugJkDBias/6D6e4Ui9ug3hXdKYmls2bN3NY75a0bFHDpIdrmDBsKwft24yX3t7IQ2813WsjRKS8ZEwM7l6VbbkxyPfittTE0qpVK55ftIm+FW0Y+ZVuTH1sGW8u28hmb8ekm9XwLCLlIVbjs5ntBvQiXAVdDSxx9w3Zn9Vw1V7cduE1l1C1+EVw2O+Ag3J+fmpiWbl5Dy69+x1O/0onBh26O53at9TwxCJSdnJKDGZ2HHAZcDTQPGHVNjN7GrjK3Z+oh/iKYo9W67nuogO2HzXk2DNpx6um9+dro87muXlPce+Vuoo6F5oARqThqXOiHjM7B/g1oWF5DvAq4WihHXAI8GXAgB+6+631Gm0G+UzUo8lpSkcTwIiU1k5N1GNmvYAbCQPkneruSzPUuQe40cwq3b2s+mRqcprS0QQwIg1TXaeSxhK6qg7LNEObuy82s2HAQuBs4OLChli/Kip68Ye/v8lzr62h6oPNVHRtxeEHdaKi4oBSh9boKSmLNEx1jZV0LHBfXdN2uvsHhEH2jitUYMXSst3e3PnI23z/sLXcP+4TRh+1gT8/9i59+x9b6tAaPU0A03hobLDGpa7EsB/hNFIuXgR65xdOcc2urOS52XdzxXd3YZ9OzXn7A6d1y61c+J3OvD7vqVKH1+hpxrjGobat6LxBK5h1eRfOG7SCaZPHKTmUsbpOJXUA1ua4rY+A9nlFU2TTp02hzS5bGdKvFS2aGwDVm7bSbPk6/vXMHI4/qm+snjLqYROP5sJoHNRW1PjUlRhaADU5bqsmh+01KFVVi9lvn92Y/84W+vcMoX+61Xnj3Q303bcD916W+8B6GlF152gujPKntqLGJ5cv8sFm1iqHekfmG0yxVVT04vCun3L5X95l1Jc30b1TDW+879z5T+OiH1TQonmznH/96FeTNBTFPnLNd2gZaXhySQxnR7dcZL8oogGZXVnJ6tWrmfTEUmpqtvLrTc3YVgO7NIdPa5on1c3l149+NUlDUIojV82b3vjUlRjKrpdRLmr/eX7+7db4V1rx/qoNXPdgDWs3Qp/PNefLBxrTH1vGkAGdgNx+/ehXkzQEpThyVVtR45M1Mbj7nGIFksjMTiJcWNcc+IO7X1fI7df+8+yxy1qWbVxP53bw4xPhjjkw7qRt3PB/23hh6RpGXfcKC9/byIYtLRh5zqlpt1V72P76a/M58/k1HHdoO9Z+/EmdzysmNYo3HaU6clVbUeMSZ87nojCz5oQhOE4G+gIjzKxvIfdRVbWYDs3XsHrFEnrtCQN6wAkHw9srYd0GOHp/2LM9nHBgNb8b25qp5+zNvCfu3KH7XWI3vWcm9WDUkPY89+oKvtIn+/OKSV0JmxZdGyKF0OASA3AYsMjdF0dzQdwNDC/kDioqevHsy+/Ss4uz267QzGDxKjioG8x4Fp55C8afBAP2a0e/focyaMA+TBjemunTpiRtJ/GwvUXzZsxfsplJP2jFF/fP/rxiSo0xnFoobUxSf3RtiBRCQ0wM+wDvJiwvi8qSmNlYM5trZnNXrVoVawcjx4znxr9vZeEKcIcXl8Kkh2HUMbBkFbzzIRzaozmbN2/+7Dn9erahqir5cDwctrfZvvxBmKinrucVU2qMUPqYpP4MGTqUMeffyNQ5e3HClauYOmcvDUoosTXE6w4sTdkOvZ2ikVxvhTC6apwdDBk6lI+2tOeqBz7m403QpT2M+jJ07QDtW8PS1fD+uhZ077q9l266w/HUBueKrmGinj06tM76vGJSo3jTo/P9kq+GeMSwDNg3Ybkb8H6hdzL2x5ewdlMLLvuWcdMZsGkLnHs7vP4+HNSrI9f8DVZu3iPr4XjqYfvhB3Xi4rtr6nxeMenUgojEVed8DMVmZi2At4AhwHvAf4DT3f21TM/Z2fkYJk2cyJ2//QUff7yOZmZ8yq502bMz/Q85hL79j+X1eU/V2ZMntcdPrs8rJvVKEpF0Ms3H0OASA4CZnQJMIXRXvc3dr8lWP5+JekREmqqdmqinVNz9EeCRUschItIUNcQ2BhERKSElBhERSaLEICIiSZQYREQkSYPslRSXma0Cqnby6Z2B1QUMp9gUf2kp/tJS/PmpcPcuqYWNIjHkw8zmpuuuVS4Uf2kp/tJS/PVDp5JERCSJEoOIiCRRYogG4itjir+0FH9pKf560OTbGEREJJmOGEREJIkSg4iIJGmyicHMTjKzN81skZldWOp4MjGz28xspZm9mlDWycweM7OF0X3HhHUXRa/pTTM7sTRRfxbLvmb2hJktMLPXzGxcVF4u8bcys+fN7OUo/iuj8rKIv5aZNTezeWb2cLRcNvGb2VIzm29mL5nZ3KisnOLf3cz+amZvRP8HR5ZF/O7e5G6E4bzfBnoBuwAvA31LHVeGWI8FBgCvJpRNBC6MHl8I/DJ63Dd6LbsCPaPX2LyEse8NDIgetyPMs9G3jOI3oG30uCXwHHBEucSf8Dp+AvwZeLicPj9RTEuBzill5RT/ncDZ0eNdgN3LIf6mesRwGLDI3Re7+xbgbmB4iWNKy92fAtakFA8nfOCI7r+RUH63u3/i7kuARYTXWhLuvtzdX4weVwMLCPN3l0v87u7ro8WW0c0pk/gBzKwb8FXgDwnFZRN/BmURv5m1J/ywmwbg7lvc/SPKIP6mmhj2Ad5NWF4WlZWLru6+HMKXL7BnVN5gX5eZ9QD6E351l0380WmYl4CVwGPuXlbxEya8ugCoSSgrp/gdmGVmL5jZ2KisXOLvBawCbo9O5f3BzNpQBvE31cRgacoaQ7/dBvm6zKwtcB8w3t0/zlY1TVlJ43f3be7+BcLc44eZ2cFZqjeo+M1sGLDS3V/I9Slpykr9+Tna3QcAJwPnmtmxWeo2tPhbEE4D3+Lu/YENhFNHmTSY+JtqYlgG7Juw3A14v0Sx7IwPzGxvgOh+ZVTe4F6XmbUkJIU/ufv9UXHZxF8rOgXwJHAS5RP/0cDXzWwp4XTp8WY2nfKJH3d/P7pfCTxAOLVSLvEvA5ZFR5kAfyUkigYff1NNDP8B9jOznma2C3Aa8GCJY4rjQWBU9HgUMDOh/DQz29XMegL7Ac+XID4AzMwI51cXuPuvElaVS/xdzGz36HFrYCjwBmUSv7tf5O7d3L0H4TP+uLuPpEziN7M2Ztau9jFwAvAqZRK/u68A3jWzA6KiIcDrlEP8pWyxL+UNOIXQS+Zt4JJSx5MlzhnAcuBTwi+KMcAewGxgYXTfKaH+JdFrehM4ucSxH0M4FH4FeCm6nVJG8R8CzIvifxW4LCovi/hTXstgtvdKKov4CefoX45ur9X+n5ZL/FE8XwDmRp+hvwEdyyF+DYkhIiJJmuqpJBERyUCJQUREkigxiIhIEiUGERFJosQgIiJJlBiaADMbbGZuZoMbQCw9oljOLHUsAmb2pJktK3Uc6ZjZmdFnpUepY2lqlBhKIOGLOtNteqljFJGmq0WpA2jiphGGWUi1uMhxFFMV0JpwwZ6INEBKDKX1b3dvUkcHHq6o3FzqOEQkM51KauDM7Iro9FI/M7vBwmxuG83s72ZWkab+QDN7ysw2mdlyM/slYYKQdNvuY2b3mdmaqP6LZvaDDPtPd1uaUvdkM3vGzDaY2cdm9qiZHZ5SJ20bQzQuzs+jma42R69zlpkdk8c+rjazERZmX9tkYSa246I6x5vZc1H5EjP7fpr3Z1czu8rM3jazT8zsXTObamYddvZvFNWZZmH2ro1m9pGZ/cPMvpRm/4MtzID3YVR3sZn9sXb8oEzMrMLMbope9/ro9pRlmRHMzPaP3sv1ZrbKzH5tYXyixDpPmtmTaZ67Q1uAmd0Rle0ZPV5rZtVmdo8lzFiWUP8EC0Nrb7Ywa9sE0ow2amZfNrM/R3+zzWa22sIMafunqetmNt3MBiX8ravM7McZ3oOvR6/x4+jzNd9SZneMPqfXJnwm3jez36R7TeVMRwyl1dbMOqcpr3b3T1LKbgPWAj8nzIz2E2A68OXaCmbWB3gC+AT4JfAxYZCuE1J3YGa9gX8R/vluJowbfxpwl5l1dffro6r3EyYMSbQ3cB3bR4XEzE4ljOD5JnAFIRn9P2COmQ1x92cyvQkWBqh7AvgScG8UTyvgKGAQ8PRO7uMU4AzgFmALMAH4PzM7A/h1VD4d+FH0ul9w9zcSnv9XYBjwF2AycGhU92gzO2pn/kbAidF27iGcVutKGP9qjpkNqN2/mR0I/J0waN/VQDXQHfga0D5azuRLwFcIg7MtBjoAI4FHzGyouz+RUr81UAnMIczdcCTwQ8IsYqdk2U8u/o8wC9vFwIHAuYS/xWc/QMxsUFRvGXAlYXytHwIfpdneqcDnCBPcvAdUAGOBf5rZwe6+KqX+IYS/3++j54wApprZ6+4+OyGGnwKTCGMyTSL8PxxImETnuqjOroSxjQ4CbiV8DvePXtMRZnZkms9EeSr1IFNN8UYY0Myz3M5MqHtFVPZAyjZ+EpX3TSj7C7ANODihrC1hUC4HBieU30uYvGVgQtkuhIl0NpEynWJCnVZRnbXAAVFZC8LwwO8CuyfU7Ub4ApubUNYjzWv836js3DT7szz2sR7YJ6H8q1H5p8ChCeUHReWTE8pOicpuSolnXFT+o538G7VJ8xo7E5Lsb9Psp8tOfL52S1O2K2EGvX+klD8Z7eeXKeWTo/KTU+o+mWbbZ0Z1eySU3RGVTUmpOxXYCrRPKPsPsA7YK6Fsr6gsdbvp3r/9CT+GLkwpd8L/w4CUz+9K4N6Uz8unwD+BVuk+f9HjCwhJbUBKna9H+zo77t+qod50Kqm0fkX4ZZd6ezRN3VtSlmt/9X0ewkxjhMlMKt391dpKHqam/F3iE6O6pxCGYZ6bUHcLYcavVqQ5yohMA74InObub0ZlAwm/kH/nYd6C2u0tI8w1/EUz+1yG7UH4FfhOmteIR/95O7mPB939vYTl2iOK59395YRtvEb4EuqdUPfr0f0vU7b5W8KR2NfZUda/UbSvDbWPzWw3M9ujNibCL/1a66L7b5pZrP9Td9+YsI9W0T7aEb7YdzhlFZmSsjw5uv9anH2n8ZuU5ScIc65XRPF1Jfxt7/YwTDXw2ZDVf0rdWMr71zZ6bWsIv97TvbbnPJpeNnr+ZsKR8ucT6nyL8MPjimh94v4SRxk9jTBS6jtm1rn2Fm1vE2FY7UZBp5JKa4G7V+ZYtypleW103ym67wK0IfyDpEotq627IE3d16P7nqkrzOxi4HTgf9w9MXn1iO7r2l6mSUf2IySpmgzrd3YfSe+Zu39kZhCSUKqP2P5e1u5vfZR4ErfxiZktJs37k7o/dvwb1c4DfDXwXcKv4kRLEh7fDZxFSOq/MLM5wCOEL9D1ZGFhcqRLCafRUtuh0g2nXO3RVJOfVXJ/38yqSf8646jrPandfi6f29qJba4jJKzU8/qrc9h/bQyHJCzvF92/kqZuoj6E026pp6tq7ZmhvOwoMZSPbRnKLeU+zjjq6eqm3Y6ZDSd8od3m7lPy3V6Oz813H5nes7rey7pisgzrctnuDMIvyxuAFwlHBjXARSQfWWyOzr0fQzgSHEo4T35p1L6ReCSUagrw34QjmKcJv6i3AaMJiT1VtteZWi/d9JPNs8RSsM9tdOQ0izAP8g2E9oD1hPdvCuk70+T6t84lhmaEo4PLMqxfm6G87CgxNB4rCXPK9kmz7oCU5VVR3b5p6tY+f2ltgZn1IzSiPkv4wklVW7cvobE66/bSWAgcbGbNshw15LuPuJYCJ5pZt8SjBgsz/vUktLPEYmE2uFOAK939ipR1V6XWj05j/DO6XWxmJxOOGs4lNOZmcjpwl7ufm7KPMRnqtzezvROPGqLTcm1JPopZS/IpmFq9ssRSl9prdnL53PYDDgZGu/sdiSvMrBPpjxhysTC6PwR4PEu9RYT2rVyP8suW2hgaCXffBvwDGGoJE9abWVvgnDR1/06YA3hAQt2WhEbPTwi/zDCzLsBDhF+d34raIVLNJcwyNzY6VVK7vc8B3wde9Gju3gzuZXvvkiQWnfspwD7ieii6n5BSfg6hV9BDxLeN8Ks06f/OwlAlqV1u92BH86L7TmnWpe4ndR8HEHrYZDI+Zfn86P7hhLKFQJ/odE7tdjsQjkR2irt/ALxAmNLys1Nr0ePULsS1v/5TX9soQvvTzro/2vblUc+jxG2nHu0daGY7HHWZWfMoOTUKOmIorSPMLN3FXh+6+993YnuXEiarf8LMbiY0kp4Z3ae6mHB6YnZUt7a76hHABHev/fX1a8KX9mTghOT/E9a7+9/cfauZjSecF/+3md3G9q6ktckmm8nAN4FbotMnT0fPP4rwZXhtAfYR1yOELpTnRQ2kTxF+sY4lnAK6Ne4G3b3azGYDF0RddN+KtjmacFok8fqES83seMIX81LCr/fRhNMmd9exqweAMWa2kfCl24twpLcA6J+m/lpgRPSF/29Cd9WRwCx3fySh3u8JCaPSzH4H7Ab8F6GbaT5fzBOAx4B/mdlvCad5ziEcTXwhod4b0W2yhetD3ick1G+Qx2gB7r7UzC4htF3MNbO7CUcf+xM+g0dGVScTjvimm9kwwmklCJ0Wvk3oXXfHzsbRoJS6W1RTvFF3d9XErpdXRGW9U7bRg5Run1H5YYRTD5sJv7B/SehhlNRdNap7IOHX0tqo/jzgjJQ6T2aJc2lK3VMIPX82ErqQzgKOyDHudoR/zLcJXQJXEo6Ajs5jH1enee8dmJ6mfCnwdErZrsBVhNMpWwj95m8iobts3L8RoYHyj2w/9fcsoc3hjsT3EzgOuI/QPfcT4IPo/RicGnua19I2ivM9Qm+ZecD3auNM8/ddRvgSfJRwzv5DQvtE2zTb/hYhwWwhnFo5l+zdVVtk+OynfhZPIiTcTwgNxhcQEmHqdj8PPEg4gq0mJJQvkKYrbZa/ddJ7nVD+neiztSF6H14GLkip04owL/OrhP+Zj6J6E4Hupf5uKdRNcz6LiEgStTGIiEgSJQYREUmixCAiIkmUGEREJIkSg4iIJFFiEBGRJEoMIiKSRIlBRESSKDGIiEiS/w/YK1iImr9aEQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "from matplotlib.pyplot import scatter,xlabel,ylabel\n",
    "\n",
    "#Extract data series from the DataFrame\n",
    "endo_data = combined_df['average_endozoicomonas_per_1000_reads']\n",
    "disease_data = combined_df['disease_susceptibility']\n",
    "\n",
    "#Convert to lists\n",
    "endo_data = list(endo_data)\n",
    "disease_data = list(disease_data)\n",
    "\n",
    "#Make a scatterplot\n",
    "scatter(endo_data,disease_data,facecolor='orange',edgecolor='black',alpha = 0.75)\n",
    "xlabel(\"Endozoicomonas abundance\", size = 'xx-large')\n",
    "ylabel(\"Disease Susceptibility\", size = 'xx-large')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5ddfba2",
   "metadata": {},
   "source": [
    "**Important Note about Statistical Independence and Comparing the Traits of Species**: This trend looks interesting, but to test for significance in this trend we would **have** to use some type of phylogenetic comparative method. This is because all 'regular' statistical methods assume data are statistically independent of each other. However, species evolve on a tree, and have traits that are correlated (NOT independent) of their relatives. This sounds like a minor issue, but actually results in **massive** false positive rates anytime you compare the traits of species without accounting for the structure of the tree. So we should treat our scatterplot as a first insight, but use something like phylogenetic independent contrasts (PIC) or phylogenetic generalized least squares (PGLS) if we wanted to test for significance."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c336461",
   "metadata": {},
   "source": [
    "## Exercises"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9949d2a8",
   "metadata": {},
   "source": [
    "- **Exercise 1**. Using the code below, make two dataframes with missing data. Then try to join them using an outer, inner, left or right join. Describe what happens in each case, paying attention to which data points are included, and which are excluded."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "id": "338de4e7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>length</th>\n",
       "      <th>n_offspring</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>individual_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               length  n_offspring\n",
       "individual_id                     \n",
       "1                10.0          NaN\n",
       "2                 NaN          1.0\n",
       "3                 5.0          3.0\n",
       "4                 7.0          5.0"
      ]
     },
     "execution_count": 212,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "#Make and examine our first set of data for individuals 1-4\n",
    "df1_data = {\"individual_id\":[1,2,3,4],\"length\":[10,None,5,7],\"n_offspring\":[None,1,3,5]}\n",
    "\n",
    "#Make a dataframe directly from our dictionary\n",
    "#NOTE: normally in real applciations we'd load our dataframe using pd.read_csv to load a file\n",
    "df1 = pd.DataFrame(df1_data)\n",
    "\n",
    "df1 = df1.set_index('individual_id')\n",
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "id": "72382ad8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>main_diet</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>individual_id</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>frogs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>newts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>frogs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>insects</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              main_diet\n",
       "individual_id          \n",
       "2                 frogs\n",
       "3                 newts\n",
       "4                 frogs\n",
       "5               insects\n",
       "6                  None"
      ]
     },
     "execution_count": 216,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Now let's say we have additional data for individuals 2-6\n",
    "df2_data = {\"individual_id\":[2,3,4,5,6], \"main_diet\":[\"frogs\",\"newts\",\"frogs\",\"insects\",None]}\n",
    "\n",
    "df2 = pd.DataFrame(df2_data)\n",
    "df2 = df2.set_index('individual_id')\n",
    "df2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2244d76",
   "metadata": {},
   "source": [
    "As a reminder, here's an example of how to join these tables:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "id": "a87bf226",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>length</th>\n",
       "      <th>n_offspring</th>\n",
       "      <th>main_diet</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>individual_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>frogs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>newts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>frogs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>insects</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               length  n_offspring main_diet\n",
       "individual_id                               \n",
       "1                10.0          NaN       NaN\n",
       "2                 NaN          1.0     frogs\n",
       "3                 5.0          3.0     newts\n",
       "4                 7.0          5.0     frogs\n",
       "5                 NaN          NaN   insects\n",
       "6                 NaN          NaN      None"
      ]
     },
     "execution_count": 220,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df = df1.join(df2,how=\"outer\")\n",
    "combined_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bfbf77b9",
   "metadata": {},
   "source": [
    "Now try changing the 'outer' in `how = 'outer'` to one of the other join methods discussed in this chapter. How do es that change the results?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37ff8bfd",
   "metadata": {},
   "source": [
    "- **Exercise 2**. Join together the data from **Exercise 1** using an 'outer' join. Next, filter that table to drop any rows with `NaN` values, using either method described earlier in the chapter."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2e2cc76",
   "metadata": {},
   "source": [
    "- **Exercise 3**. Imagine that you wanted to compare the `main_diet` and `n_offspring` columns to see if diet affected the number of offspring your organisms had.  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08a6296c",
   "metadata": {},
   "source": [
    "## Reading Responses & Feedback \n",
    "\n",
    "^Make this a hyperlink using Google Forms"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d74d3ba",
   "metadata": {},
   "source": [
    "## Further Reading\n",
    " - [Selecting Subsets of Data in Pandas Part 1](https://www.dunderdata.com/blog/selecting-subsets-of-data-in-pandas-part-1)\n",
    " - [Selecting Subsets of Data in Pandas Part 2](https://www.dunderdata.com/blog/selecting-subsets-of-data-in-pandas-part-2)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
