{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Bioinformatics Vignette: Using For Loops and Dictionaries to Compare Nucleotide Composition in Pandemic and Non-Pandemic Causing Influenza Strains"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## *Vignette Author*: Nia Prabhu"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "<img src=\"./nia_prabhu_vignette_picture.png\" width=\"200\" align=\"left\" style=\"margin: 0px 10px 0px 0px;\" description=\"A professional picture of Nia Prabhu, the author of this vignette\"> \n",
    "\n",
    "## About Me \n",
    "I am passionate about medicine and research! Bioinformatics intrigues me because it blends together the diverse fields of computer science, biology, mathematics, and physics. The beauty of bioinformatics is that it uses computational methods to take large and complex data sets like the human genome and output very specific biological data which we can then use to test a hypothesis. Something that I would like to further study in this field is the analysis of data from genome sequencing and gene variants in a patient’s response to a particular drug. \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Background\n",
    "\n",
    "Influenza A pandemics occur when the human population is exposed to a new Influenza A virus strain that it has little or no pre-existing immunity to. Antigenic shifts in Influenza A viruses result in a new HA or HA/NA protein subtype, which then gains the ability to infect humans. This novel subtype can cause an influenza pandemic. \n",
    "\n",
    "These shifts transpire when an influenza virus from an animal (most commonly from a swine or avian origin) infects humans with the novel subtype, and due to the lack of immunity to this, the disease rapidly transmits among the population. Given the detrimental nature of influenza viruses and the several adverse impacts they’ve brought, it’s important to study these viruses and their mutations in order to forecast for future outbreaks of disease. \n",
    "\n",
    "My research focused on two specific glycoproteins on the surface of Influenza A viruses - HA and NA- in both pandemic and non-pandemic strains. I analyzed nucleotide composition to study the similarities and patterns between the nucleotides of these two proteins in pandemic or non-pandemic strains. This analysis was possible because other researchers had already used genetic characterization to study many influenza viruses and share their sequences in online databases. \n",
    "\n",
    "Genetic characterization is the process of comparing genetic sequences, and genomic sequencing is a type of genetic characterization in which we can determine the nucleotides present in each gene of the virus’s genome. Unlike our double-stranded DNA genome, Influenza A's genome is an RNA genome (A,U,C, and G) that is single-stranded and broken up into pieces called segments. However, this segmented RNA genome is usually converted into DNA using a reverse transcriptase protein before being studied with DNA sequencing, so our data was still presented as DNA (A,T,C, and G). \n",
    "\n",
    "Something I was specifically interested in was studying the nucleotide compositions and comparing the pandemic H1N1 strains to non-pandemic H1N1 strains to determine patterns among Influenza A viruses. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Applying the Technique\n",
    "\n",
    "### Data Collection\n",
    "The database that we used to collect our genomic sequences was the [Influenza Research Database]( https://www.fludb.org/brc/influenza_sequence_search_segment_display.spg?method=ShowCleanSearch&decorator=influenza). For data collection, I used a dataset of HA and NA protein segments from pandemic H1N1 reference strains and non-pandemic H1N1 reference strains. The dataset consisted of a total of 40 sequences- 10 HA segments from pandemic strains, 10 NA segments from pandemic strains, 10 HA segments from non-pandemic strains, and 10 NA segments from non-pandemic strains. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Table 1. HA and NA nucleotide sequences from pandemic and non-pandemic causing strains.**\n",
    "\n",
    "| Link to Database                                                                                                        \t| Protein Type \t| Pandemic causing? \t| Strain Name                            \t|   \t|\n",
    "|-------------------------------------------------------------------------------------------------------------------------\t|--------------\t|-------------------\t|----------------------------------------\t|---\t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=GQ280797&decorator=influenza&context=1628317532186 \t| HA           \t| yes               \t| A/California/04/2009                   \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=FJ969517&decorator=influenza&context=1628317555652 \t| NA           \t| yes               \t| A/California/04/2009                   \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=HM138501&decorator=influenza&context=1628317639251 \t| HA           \t| yes               \t| A/Germany-BY/74/2009 A/AA/Marton/1943  \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=HM138500&decorator=influenza&context=1628317679677 \t| NA           \t| yes               \t| A/Germany-BY/74/2009 A/AA/Marton/1943  \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=CY040888&decorator=influenza&context=1628317752983 \t| HA           \t| yes               \t| A/Mexico/47N/2009(H1N1)                \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=CY040890&decorator=influenza&context=1628317817620 \t| NA           \t| yes               \t| A/Mexico/47N/2009(H1N1)                \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=GQ402188&decorator=influenza&context=1628317850766 \t| HA           \t| yes               \t| A/Mexico/InDRE4115/2009                \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=GQ402228&decorator=influenza&context=1628317893386 \t| NA           \t| yes               \t| A/Mexico/InDRE4115/2009                \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=GQ894830&decorator=influenza&context=1628317943477 \t| HA           \t| yes               \t| A/Rhode Island/08/2009                 \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=GQ894831&decorator=influenza&context=1628317971411 \t| NA           \t| yes               \t| A/Rhode Island/08/2009                 \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=CY053744&decorator=influenza&context=1628318065357 \t| HA           \t| yes               \t| A/Russia/190/2009                      \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=CY053746&decorator=influenza&context=1628318084753 \t| NA           \t| yes               \t| A/Russia/190/2009                      \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=CY071063&decorator=influenza&context=1628318131697 \t| HA           \t| yes               \t| A/Tallinn/INS372/2009                  \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=CY071065&decorator=influenza&context=1628318173047 \t| NA           \t| yes               \t| A/Tallinn/INS372/2009                  \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=CY074982&decorator=influenza&context=1628318303335 \t| HA           \t| yes               \t| A/Thailand/CU-H1222/2010               \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=CY074984&decorator=influenza&context=1628318830367 \t| NA           \t| yes               \t| A/Thailand/CU-H1222/2010               \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=CY080420&decorator=influenza&context=1628319077054 \t| HA           \t| yes               \t| A/Ulaanbaatar/190/2011                 \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=CY080574&decorator=influenza&context=1628319197987 \t| NA           \t| yes               \t| A/Ulaanbaatar/190/2011                 \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=GQ397278&decorator=influenza&context=1628319283045 \t| HA           \t| yes               \t| A/Yamaguchi/22/2009                    \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=GQ397279&decorator=influenza&context=1628319315358 \t| NA           \t| yes               \t| A/Yamaguchi/22/2009                    \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=CY021709&decorator=influenza&context=1628320312832 \t| HA           \t| no                \t| A/AA/Huston/1945                       \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=CY021711&decorator=influenza&context=1628320337871 \t| NA           \t| no                \t| A/AA/Huston/1945                       \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=CY020285&decorator=influenza&context=1628320421148 \t| HA           \t| no                \t| A/AA/Marton/1943                       \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=CY020285&decorator=influenza&context=1628320461651 \t| NA           \t| no                \t| A/AA/Marton/1943                       \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=CY021901&decorator=influenza&context=1628320507754 \t| HA           \t| no                \t| A/Albany/1618/1951                     \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=CY021903&decorator=influenza&context=1628320551208 \t| NA           \t| no                \t| A/Albany/1618/1951                     \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=CY026411&decorator=influenza&context=1628320634213 \t| HA           \t| no                \t| A/Albany/8/1979                        \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=CY026413&decorator=influenza&context=1628320694306 \t| NA           \t| no                \t| A/Albany/8/1979                        \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=CY074243&decorator=influenza&context=1628320744651 \t| HA           \t| no                \t| A/California/VRDL225/2009              \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=CY074245&decorator=influenza&context=1628320777620 \t| NA           \t| no                \t| A/California/VRDL225/2009              \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=CY010260&decorator=influenza&context=1628320807968 \t| HA           \t| no                \t| A/Canterbury/48/2001                   \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=CY010262&decorator=influenza&context=1628320840500 \t| NA           \t| no                \t| A/Canterbury/48/2001                   \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=AJ289702&decorator=influenza&context=1628320921483 \t| HA           \t| no                \t| A/Fiji/15899/83                        \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=AJ006954&decorator=influenza&context=1628320944073 \t| NA           \t| no                \t| A/Fiji/15899/83                        \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=U02085&decorator=influenza&context=1628321015721   \t| HA           \t| no                \t| A/Fort Monmouth/1/1947                 \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=CY009614&decorator=influenza&context=1628321060539 \t| NA           \t| no                \t| A/Fort Monmouth/1/1947                 \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=CY020461&decorator=influenza&context=1628321098549 \t| HA           \t| no                \t| A/Iowa/1943                            \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=CY020463&decorator=influenza&context=1628321120028 \t| NA           \t| no                \t| A/Iowa/1943                            \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=CY019971&decorator=influenza&context=1628321172939 \t| HA           \t| no                \t| A/Roma/1949                            \t|   \t|\n",
    "| https://www.fludb.org/brc/fluSegmentDetails.spg?ncbiGenomicAccession=CY019973&decorator=influenza&context=1628321192613 \t| NA           \t| no                \t| A/Roma/1949                            \t|   \t|"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sequence Data Analysis\n",
    "\n",
    "I compared the sequences in two ways: by comparing the nucleotide composition of each sequence, and by comparing the sequence similarity of each pair of sequences. This vignette includes code for calculating nucleotide composition of each sequence.\n",
    "\n",
    "This was done by parsing FASTA files, using for loops and if statements to find identifier lines and actual sequences. Nucleotide composition of each sequence was calculated using for loops and a dictionary to count the number of each nucleotide present in the sequence. Finally I calculated the nucleotide composition percentage.\n",
    "\n",
    "Separately, additional code was used to test how similar each pair of sequences was with other sequences in the analysis (not included in this vignette).\n",
    "\n",
    "The code used to calculate nucleotide compositions is added below. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Code \n",
    "\n",
    "I wrote python code to perform this analysis with my project partner Shubham Kamboj. For the nucleotide composition analysis, I defined two functions. The first function is called `writeFileToDict`. It takes the location of a FASTA file, along with a dictionary. It parses the FASTA file and adds the DNA sequences in it into the `dict` called `myDict`. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    " def writeFileToDict(myFile, myDict):\n",
    "    \"\"\"Function to write the file contents to a dictionary\n",
    "    \"\"\"\n",
    "    seqID = None\n",
    "    sequence = []\n",
    "    infile = open(myFile)\n",
    "    for line in infile: \n",
    "        line = line.strip() #strip the line of any unnecessary content \n",
    "        if line.startswith(\">\"): #if it is an identifier line \n",
    "            if seqID is not None:  #if exist, put it in dictionary \n",
    "                myDict[seqID] = ''.join(sequence) \n",
    "                \n",
    "            #Set each ID to the filename plus the FASTA label line    \n",
    "            seqID = str(myFile) + \"_\" + line[1:]\n",
    "            \n",
    "            sequence = []\n",
    "        \n",
    "            continue \n",
    "                \n",
    "        sequence.append(line) #append sequence \n",
    "            \n",
    "    myDict[seqID] = ''.join(sequence)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The second function, `findPercentNucleotides` calculates the percentage of nucleotides in the sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def findPercentNucleotides(sequence):\n",
    "    \"\"\"Finds the percentage of nucleotides in a sequence\n",
    "    \"\"\" \n",
    "    nucleotides = [\"A\", \"G\", \"T\", \"C\"] #nucleotides \n",
    "    nucleotidePercent = {} #corresponding dictionary \n",
    "    length = len(sequence)\n",
    "    for nuc in nucleotides:\n",
    "        count = sequence.count(nuc)\n",
    "        nucPercent = (count/length)*100\n",
    "        nucleotidePercent[nuc] = nucPercent\n",
    "    return nucleotidePercent\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we can use these functions to calculate our results. For this vignette we'll run the code on a test example where the non-Pandemic sequences are all 'A' or 'T' nucleotides, and the Pandemic sequences are all 'G' or 'C' nucleotides."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./PandemicStrains/test_pandemic_strain1.txt_gene1_all_A {'A': 100.0, 'G': 0.0, 'T': 0.0, 'C': 0.0}\n",
      "./PandemicStrains/test_pandemic_strain1.txt_gene2_all_T {'A': 0.0, 'G': 0.0, 'T': 100.0, 'C': 0.0}\n",
      "./PandemicStrains/test_pandemic_strain2.txt_gene1_all_A_strain2 {'A': 100.0, 'G': 0.0, 'T': 0.0, 'C': 0.0}\n",
      "./PandemicStrains/test_pandemic_strain2.txt_gene2_all_T_strain2 {'A': 0.0, 'G': 0.0, 'T': 100.0, 'C': 0.0}\n",
      "./NonPandemicStrains/test_nonpandemic_strain1.txt_gene1_all_A {'A': 0.0, 'G': 100.0, 'T': 0.0, 'C': 0.0}\n",
      "./NonPandemicStrains/test_nonpandemic_strain1.txt_gene2_all_T {'A': 0.0, 'G': 0.0, 'T': 0.0, 'C': 100.0}\n",
      "./NonPandemicStrains/test_nonpandemic_strain2.txt_gene1_all_A_strain2 {'A': 0.0, 'G': 100.0, 'T': 0.0, 'C': 0.0}\n",
      "./NonPandemicStrains/test_nonpandemic_strain2.txt_gene2_all_T_strain2 {'A': 0.0, 'G': 0.0, 'T': 0.0, 'C': 100.0}\n"
     ]
    }
   ],
   "source": [
    "#load external tools we'll need to find and access files from our directory \n",
    "from os import listdir\n",
    "from os.path import join\n",
    "\n",
    "nonPandemicDict = {}\n",
    "PandemicDict = {}\n",
    "\n",
    "nonPandDir = \"NonPandemicStrains\"\n",
    "\n",
    "#accesses the nonpandemic strains and places them in our python nonPandemicDict dictionary \n",
    "for file in listdir(nonPandDir):\n",
    "    writeFileToDict(\"./\" + nonPandDir + \"/\" + file, nonPandemicDict)\n",
    "    \n",
    "PandDir = \"PandemicStrains\"\n",
    "\n",
    "#accesses the pandemic strains and places them in our python PandemicDict dictionary \n",
    "for file in listdir(PandDir):\n",
    "    writeFileToDict(\"./\" + PandDir + \"/\" + file, PandemicDict)\n",
    "\n",
    "#finds the percent nucleotides in each sequence with their corresponding ID \n",
    "for seqID,seq in PandemicList.items(): \n",
    "    print(seqID, str(findPercentNucleotides(seq)))\n",
    "    \n",
    "for seqID, seq in nonPandemicList.items(): \n",
    "    print(seqID, str(findPercentNucleotides(seq)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The output of the script lists each gene in each strain, followed by the percentage of each nucleotide in that gene."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercises\n",
    "\n",
    "1. The `writeFileToDict` function doesn't return a value like a traditional python function. Yet in this example, we're still getting data out of the file. Carefully read the code for this function, and see if you can figure out how data is getting *out*. *Hint*: mutable objects such as lists and dicts that are passed to a function using its arguments can be changed inside functions and those changes will be preserved even after the function ends (see https://www.dataquest.io/blog/tutorial-functions-modify-lists-dictionaries-python/ for examples, and https://realpython.com/python-pass-by-reference/ for a detailed discussion contrasting python with C#)\n",
    "2. Try modifying the `writeFileToDict` function to pass data out in a dictionary using the `return` keyword.\n",
    "3. Imagine that you liked this approach, and wanted to do something similar. However, you noticed that your DNA sequences were all represented in lowercase, whereas the sequences that Nia was studying were all uppercase. Do you think the results of running this code would be the same if the DNA sequence files you used had sequences in lowercase (e.g. like 'atgtacgatcgtagc')? Why or why not? *Hint*: you can try out example sequences that you come up with by passing them to the the findPercentNucleotides function (e.g. `findPercentNucleotides('ATAA')`)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [Reading Responses & Feedback](https://docs.google.com/forms/d/e/1FAIpQLSeUQPI_JbyKcX1juAFLt5z1CLzC2vTqaCYySUAYCNElNwZqqQ/viewform?usp=pp_url&entry.2118603224=Bioinformatics+Vignette+-+Nia+Prabhu+-+Using+For+Loops+and+Dictionaries+to+Compare+Nucleotide+Composition+in+Pandemic+and+Non-Pandemic+Causing+Influenza+Strains)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## References\n",
    "\n",
    "1.\tAl-Muharrmi, Z. Understanding the Influenza A H1N1 2009 Pandemic. *Sultan Qaboos Univ. Med. J.* 10, 187–195 (2010).\n",
    "\n",
    "2.\tJilani, T. N., Jamil, R. T. & Siddiqui, A. H.. H1N1 Influenza. in *StatPearls* (StatPearls Publishing, 2021)."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
